### PR TITLE
Fix week number for date_format evalengine function

### LINKED ISSUE #17431: Bug Report: Incorrect result for Date_format for a week of the year
### Overview of the Issue

It was noticed that the evalengine test was failing https://github.com/vitessio/vitess/actions/runs/12498442377/job/34872825175?pr=17417. The cause turned out to be the following query - 

```mysql
mysql> SELECT DATE_FORMAT(time '130:34:58', "%u");
+-------------------------------------------------------+
| DATE_FORMAT(CAST(:vtg1 AS TIME), :vtg2 /* VARCHAR */) |
+-------------------------------------------------------+
| 01                                                    |
+-------------------------------------------------------+
1 row in set (0.01 sec)

mysql> SELECT DATE_FORMAT(timestamp '2024-12-30 10:34:58', "%u");
+-----------------------------------------------------------+
| DATE_FORMAT(CAST(:vtg1 AS DATETIME), :vtg2 /* VARCHAR */) |
+-----------------------------------------------------------+
| 01                                                        |
+-----------------------------------------------------------+
1 row in set (0.00 sec)
```

These are the outputs from vtgate, while MySQL outputs 53, which also intuitively makes sense. 30th December lies in the 53rd week and not the first week :smile

### Reproduction Steps

1. Run vtgate
2. Run the queries listed above

### Binary Version

```sh
main
```

### Operating System and Environment details

```sh
-
```

### Log Fragments

_No response_

### PR DESCRIPTION
**Pull request by GuptaManan100**
## Description

This PR fixes the way evalengine finds the week for the date_format function. It was noticed that the `Week()` function works fine, but we weren't using it. This PR changes that to fix the issue ran into in #17431.

## Related Issue(s)

- Fixes #17431 

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/mysql/datetime/spec.go [hunk 1]
```diff
@@ -374,10 +374,7 @@ func (u fmtWeek0) parse(t *timeparts, bytes string) (string, bool) {
 type fmtWeek1 struct{}
 
 func (fmtWeek1) format(dst []byte, t DateTime, prec uint8) []byte {
-	year, week := t.Date.ISOWeek()
-	if year < t.Date.Year() {
-		week = 0
-	}
+	week := t.Date.Week(1)
```
Comment (GuptaManan100) [inline go/mysql/datetime/spec.go:377]: I can make the same change for the other `fmtWeek` functions. Haven't done that yet, but i think it makes sense to just use the same implementation for everything.

**Comment by GuptaManan100** [comment]
I'm not sure about the backports either. Do we need to backport this fix to older releases?

File: go/mysql/datetime/spec.go (hunk 1 above)
Comment (dbussink) [reply go/mysql/datetime/spec.go:377]: @GuptaManan100 i think we should do that indeed. Makes it less error prone and the week function afaik is already better tested too.

**Review by harshit-gangal — APPROVED** [review]

File: go/mysql/datetime/spec.go (hunk 1 above)
Comment (dbussink) [reply go/mysql/datetime/spec.go:374]: @GuptaManan100 Pushed up the fix to always used the documented mode number.

**Comment by dbussink** [comment]
> I'm not sure about the backports either. Do we need to backport this fix to older releases?

Don't have a super strong opinion, we could back port to v21 only as a small bugfix?

**Review by dbussink — APPROVED** [review]

**Pull request merged**